File: LeetCode/python3/200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dir_x, dir_y = [0, 1, 0, -1], [1, 0, -1, 0]
        num, m, n = 0, len(grid), len(grid[0]) if len(grid) > 0 else 0
        if m == 0:
            return 0

        def DFS(x: int, y: int):
            if grid[x][y] == '0':
                return
            grid[x][y] = '0'
            for k in range(4):
                new_x, new_y = x + dir_x[k], y + dir_y[k]
                if new_x >= 0 and new_x < m and new_y >= 0 and new_y < n:
                    DFS(new_x, new_y)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    DFS(i, j)
                    num += 1
        return num

File: LeetCode/cpp/200.cpp
class Solution {
    vector<vector<char>> g;
    int dir_x[4] = {-1, 0, 1, 0};
    int dir_y[4] = {0, -1, 0, 1};
public:
    int numIslands(vector<vector<char>> &grid) {
        int m = grid.size(), n = m != 0 ? grid[0].size() : 0;
        if (m == 0)
            return 0;
        int num = 0;
        g = grid;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (g[i][j] == '1') {
                    FindRoot(i, j);
                    ++num;
                }
            }
        }
        return num;
    }

    void FindRoot(int i, int j) {
        g[i][j] = '2';
        for (int l = 0; l < 4; ++l) {
            int x = i + dir_x[l], y = j + dir_y[l];
            if (x >= 0 && y >= 0 && x < g.size() && y < g[0].size() && g[x][y] == '1') {
                FindRoot(x, y);
            }
        }
    }
};